lms/djangoapps/program_enrollments/api/v1/views.py

```diff
@@ -1209,3 +1204,188 @@ def get_course_run_status(course_overview, certificate_info):
+        grade_results = self._load_grade_results(program_uuid, course_key)
+        serializer = ProgramCourseGradeResultSerializer(grade_results, many=True)
+        response_code = self._calc_response_code(grade_results)
+        return self.get_paginated_response(serializer.data, status_code=response_code)
+
+    def _load_grade_results(self, program_uuid, course_key):
+        """
+        Load grades (or grading errors) for a given program courserun.
+
+        Arguments:
+            program_uuid (str)
+            course_key (CourseKey)
+
+        Returns: list[ProgramCourseGradeResult|ProgramCourseGradeErrorResult]
+        """
+        enrollments_qs = use_read_replica_if_available(
+            ProgramCourseEnrollment.objects.filter(
+                program_enrollment__program_uuid=program_uuid,
+                program_enrollment__user__isnull=False,
+                course_key=course_key,
+            ).select_related(
+                'program_enrollment',
+                'program_enrollment__user',
+            )
+        )
+        paginated_enrollments = self.paginate_queryset(enrollments_qs)
+        if not paginated_enrollments:
+            return []
+
+        # Hint: `zip(*(list))` can be read as "unzip(list)"
+        enrollments, users = zip(*(
+            (enrollment, enrollment.program_enrollment.user)
+            for enrollment in paginated_enrollments
+        ))
+        enrollment_grade_pairs = zip(
+            enrollments, self._iter_grades(course_key, list(users))
+        )
+        grade_results = [
+            (
+                ProgramCourseGradeResult(enrollment, grade)
+                if grade
+                else ProgramCourseGradeErrorResult(enrollment, exception)
+            )
+            for enrollment, (grade, exception) in enrollment_grade_pairs
+        ]
+        return grade_results
+
+    @staticmethod
+    def _iter_grades(course_key, users):
+        """
+        Load a user grades for a course, using bulk fetching for efficiency.
+
+        Arguments:
+            course_key (CourseKey)
+            users (list[User])
+
+        Returns: iterable[( CourseGradeBase|NoneType, Exception|NoneType )]
+            Iterable of pairs, in same order as `users`.
+            The first item in the pair is the grade, or None if loading the
+                grade failed.
+            The second item in the pair is an exception or None.
+        """
+        prefetch_course_grades(course_key, users)
+        try:
+            grades_iter = CourseGradeFactory().iter(users, course_key=course_key)
+            for user, course_grade, exception in grades_iter:
+                if not course_grade:
+                    fmt = 'Failed to load course grade for user ID {} in {}.\n{}'
+                    err_str = fmt.format(
+                        user.id,
+                        course_key,
+                        text_type(exception) if exception else ''
```

Comment:
(to critique my own code for a moment): Might be better to have this log `'Failed to load course grade for user ID {} in {}: {}"`, defaulting to `'Unknown error'` if `exception is None`. That way, someone reading the logs knows that the exception was unavailable, as opposed to us just not logging it.

Reply:
👍